src/test/unit_test/multi_runner.cpp

```diff
@@ -0,0 +1,545 @@
+
+    // combine the two top collections
+    boost::container::static_vector<run_time, 2 * max_top> top_result;
+    top_result.resize(top.size() + r.top.size());
+    std::merge(
+        top.begin(),
+        top.end(),
+        r.top.begin(),
+        r.top.end(),
+        top_result.begin(),
+        [](run_time const& t1, run_time const& t2) {
+            return t1.second > t2.second;
+        });
+
+    if (top_result.size() > max_top)
+        top_result.resize(max_top);
+
+    top = top_result;
+}
+
+template <class S>
+void
+results::print(S& s)
+{
+    using namespace beast::unit_test;
+
+    if (top.size() > 0)
+    {
+        s << "Longest suite times:\n";
+        for (auto const& i : top)
+            s << std::setw(8) << fmtdur(i.second) << " " << i.first << '\n';
+    }
+
+    auto const elapsed = clock_type::now() - start;
+    s << fmtdur(elapsed) << ", " << amount{suites, "suite"} << ", "
+      << amount{cases, "case"} << ", " << amount{total, "test"} << " total, "
+      << amount{failed, "failure"} << std::endl;
+}
+
+//------------------------------------------------------------------------------
+
+template <bool IsParent>
+std::size_t
+multi_runner_base<IsParent>::inner::checkout_job_index()
+{
+    return job_index_++;
+}
+
+template <bool IsParent>
+std::size_t
+multi_runner_base<IsParent>::inner::checkout_test_index()
+{
+    return test_index_++;
+}
+
+template <bool IsParent>
+bool
+multi_runner_base<IsParent>::inner::any_failed() const
+{
+    return any_failed_;
+}
+
+template <bool IsParent>
+void
+multi_runner_base<IsParent>::inner::any_failed(bool v)
+{
+    any_failed_ = any_failed_ || v;
+}
+
+template <bool IsParent>
+void
+multi_runner_base<IsParent>::inner::inc_keep_alive_count()
+{
+    ++keep_alive_;
+}
+
+template <bool IsParent>
+std::size_t
+multi_runner_base<IsParent>::inner::get_keep_alive_count()
+{
+    return keep_alive_;
+}
+
+template <bool IsParent>
+void
+multi_runner_base<IsParent>::inner::add(results const& r)
+{
+    std::lock_guard<boost::interprocess::interprocess_mutex> l{m_};
+    results_.merge(r);
+}
+
+template <bool IsParent>
+template <class S>
+void
+multi_runner_base<IsParent>::inner::print_results(S& s)
+{
+    std::lock_guard<boost::interprocess::interprocess_mutex> l{m_};
+    results_.print(s);
+}
+
+template <bool IsParent>
+multi_runner_base<IsParent>::multi_runner_base()
+{
+    try
+    {
+        if (IsParent)
+        {
+            // cleanup any leftover state for any previous failed runs
+            boost::interprocess::shared_memory_object::remove(shared_mem_name_);
+            boost::interprocess::message_queue::remove(message_queue_name_);
+        }
+
+        shared_mem_ = boost::interprocess::shared_memory_object{
+            std::conditional_t<
+                IsParent,
+                boost::interprocess::create_only_t,
+                boost::interprocess::open_only_t>{},
+            shared_mem_name_,
+            boost::interprocess::read_write};
```

Comment:
I launched a `--unittest-child` process without a parent running, and it raised an exception here. It's pretty obvious that this violates the warranty, but do we want to handle more gracefully than this?

```
$ ./build/cmake/msvc/Debug/rippled --unittest --unittest-jobs 5 --unittest-child
Terminating thread rippled: main: unhandled class boost::interprocess::interprocess_exception 'The system cannot find the file specified.'
```

Reply:
I don't think that's worth worrying about. It throws an exception (which is a good thing here), and it's clearly being misused.